File: benchmarks/rust/cpu_benchmark.rs
// CPU Benchmark - Rust
// Tests: Fibonacci, Matrix Multiply, Sorting, Prime Sieve

use std::time::Instant;

// ─── Fibonacci (Recursive) ──────────────────────────────────────────────────

fn fib_recursive(n: u64) -> u64 {
    if n <= 1 {
        return n;
    }
    fib_recursive(n - 1) + fib_recursive(n - 2)
}

fn benchmark_fibonacci() -> f64 {
    let start = Instant::now();
    
    let mut sum = 0u64;
    for i in 0..40 {
        sum += fib_recursive(i);
    }
    
    let elapsed = start.elapsed().as_millis() as f64;
    println!("Fibonacci(40): {} ms", elapsed);
    elapsed
}

// ─── Matrix Multiply ────────────────────────────────────────────────────────

fn matrix_multiply(a: &[Vec<f64>], b: &[Vec<f64>], c: &mut [Vec<f64>], n: usize) {
    for i in 0..n {
        for j in 0..n {
            c[i][j] = 0.0;
            for k in 0..n {
                c[i][j] += a[i][k] * b[k][j];
            }
        }
    }
}

fn benchmark_matrix_multiply() -> f64 {
    let n = 512;
    let mut a = vec![vec![0.0; n]; n];
    let mut b = vec![vec![0.0; n]; n];
    let mut c = vec![vec![0.0; n]; n];
    
    // Initialize matrices
    for i in 0..n {
        for j in 0..n {
            a[i][j] = (i * j) as f64;
            b[i][j] = (i + j) as f64;
        }
    }
    
    let start = Instant::now();
    matrix_multiply(&a, &b, &mut c, n);
    let elapsed = start.elapsed().as_millis() as f64;
    
    println!("Matrix Multiply (512x512): {} ms", elapsed);
    elapsed
}

// ─── Sorting (Quicksort) ────────────────────────────────────────────────────

fn quicksort(arr: &mut [i64], low: isize, high: isize) {
    if low < high {
        let pivot = partition(arr, low, high);
        quicksort(arr, low, pivot - 1);
        quicksort(arr, pivot + 1, high);
    }
}

fn partition(arr: &mut [i64], low: isize, high: isize) -> isize {
    let pivot = arr[high as usize];
    let mut i = low - 1;
    
    for j in low..high {
        if arr[j as usize] < pivot {
            i += 1;
            arr.swap(i as usize, j as usize);
        }
    }
    arr.swap((i + 1) as usize, high as usize);
    i + 1
}

fn benchmark_sorting() -> f64 {
    let n = 1_000_000;
    let mut arr: Vec<i64> = (0..n).rev().collect();
    
    let start = Instant::now();
    quicksort(&mut arr, 0, (n - 1) as isize);
    let elapsed = start.elapsed().as_millis() as f64;
    
    println!("Sorting (1M elements): {} ms", elapsed);
    elapsed
}

// ─── Prime Sieve (Sieve of Eratosthenes) ────────────────────────────────────

fn sieve_of_eratosthenes(limit: usize) -> usize {
    let mut sieve = vec![true; limit + 1];
    sieve[0] = false;
    sieve[1] = false;
    
    let mut p = 2;
    while p * p <= limit {
        if sieve[p] {
            let mut i = p * p;
            while i <= limit {
                sieve[i] = false;
                i += p;
            }
        }
        p += 1;
    }
    
    sieve.iter().filter(|&&x| x).count()
}

fn benchmark_prime_sieve() -> f64 {
    let limit = 10_000_000;
    
    let start = Instant::now();
    let count = sieve_of_eratosthenes(limit);
    let elapsed = start.elapsed().as_millis() as f64;
    
    println!("Prime Sieve ({}): {} primes in {} ms", limit, count, elapsed);
    elapsed
}

// ─── String Processing ──────────────────────────────────────────────────────

fn string_concatenation(n: usize) -> String {
    let mut result = String::new();
    for i in 0..n {
        result.push_str(&format!("Item {}: ", i));
    }
    result
}

fn benchmark_string_processing() -> f64 {
    let n = 100_000;
    
    let start = Instant::now();
    let _result = string_concatenation(n);
    let elapsed = start.elapsed().as_millis() as f64;
    
    println!("String Concatenation ({}): {} ms", n, elapsed);
    elapsed
}

// ─── Main ───────────────────────────────────────────────────────────────────

fn main() {
    println!("=== Rust CPU Benchmark ===");
    println!();
    
    let mut total = 0.0;
    
    total += benchmark_fibonacci();
    total += benchmark_matrix_multiply();
    total += benchmark_sorting();
    total += benchmark_prime_sieve();
    total += benchmark_string_processing();
    
    println!();
    println!("Total: {} ms", total);
}

File: benchmarks/python/cpu_benchmark.py
#!/usr/bin/env python3
"""
CPU Benchmark - Python
Tests: Fibonacci, Matrix Multiply, Sorting, Prime Sieve
"""

import time
import math


# ─── Fibonacci (Recursive) ──────────────────────────────────────────────────

def fib_recursive(n):
    if n <= 1:
        return n
    return fib_recursive(n - 1) + fib_recursive(n - 2)


def benchmark_fibonacci():
    start = time.time()
    total = 0
    for i in range(40):
        total += fib_recursive(i)
    elapsed = (time.time() - start) * 1000
    print(f"Fibonacci(40): {elapsed:.2f} ms")
    return elapsed


# ─── Matrix Multiply ────────────────────────────────────────────────────────

def matrix_multiply(a, b, c, n):
    for i in range(n):
        for j in range(n):
            c[i][j] = 0.0
            for k in range(n):
                c[i][j] += a[i][k] * b[k][j]


def benchmark_matrix_multiply():
    n = 512
    a = [[float(i * j) for j in range(n)] for i in range(n)]
    b = [[float(i + j) for j in range(n)] for i in range(n)]
    c = [[0.0] * n for _ in range(n)]

    start = time.time()
    matrix_multiply(a, b, c, n)
    elapsed = (time.time() - start) * 1000
    print(f"Matrix Multiply (512x512): {elapsed:.2f} ms")
    return elapsed


# ─── Sorting (Quicksort) ────────────────────────────────────────────────────

def quicksort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quicksort(arr, low, pi - 1)
        quicksort(arr, pi + 1, high)


def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def benchmark_sorting():
    n = 1_000_000
    arr = list(range(n, 0, -1))

    start = time.time()
    quicksort(arr, 0, n - 1)
    elapsed = (time.time() - start) * 1000
    print(f"Sorting (1M elements): {elapsed:.2f} ms")
    return elapsed


# ─── Prime Sieve (Sieve of Eratosthenes) ────────────────────────────────────

def sieve_of_eratosthenes(limit):
    sieve = [True] * (limit + 1)
    sieve[0] = False
    sieve[1] = False

    p = 2
    while p * p <= limit:
        if sieve[p]:
            for i in range(p * p, limit + 1, p):
                sieve[i] = False
        p += 1

    return sum(sieve)


def benchmark_prime_sieve():
    limit = 10_000_000

    start = time.time()
    count = sieve_of_eratosthenes(limit)
    elapsed = (time.time() - start) * 1000
    print(f"Prime Sieve ({limit}): {count} primes in {elapsed:.2f} ms")
    return elapsed


# ─── String Processing ──────────────────────────────────────────────────────

def string_concatenation(n):
    parts = []
    for i in range(n):
        parts.append(f"Item {i} ")
    return "".join(parts)


def benchmark_string_processing():
    n = 100_000

    start = time.time()
    _result = string_concatenation(n)
    elapsed = (time.time() - start) * 1000
    print(f"String Concatenation ({n}): {elapsed:.2f} ms")
    return elapsed


# ─── Main ───────────────────────────────────────────────────────────────────

def main():
    print("=== Python CPU Benchmark ===")
    print()

    total = 0.0

    total += benchmark_fibonacci()
    total += benchmark_matrix_multiply()
    total += benchmark_sorting()
    total += benchmark_prime_sieve()
    total += benchmark_string_processing()

    print()
    print(f"Total: {total:.2f} ms")


if __name__ == "__main__":
    main()
